### PR TITLE
Make JetStream consumer deliver policy configurable

### PR DESCRIPTION
**Pull request by marcobebway**
**Description**

Make JetStream consumer deliver policy configurable as per JetStream [docs](https://docs.nats.io/nats-concepts/jetstream/consumers#deliverpolicy-optstartseq-optstarttime).

**Note**

The deliver policy is set to "new" in the Eventing charts.

### CODE REVIEW

File: resources/eventing/values.yaml [hunk 1]
```diff
@@ -23,6 +23,21 @@ global:
     enabled: false
     # Storage type of the stream, memory or file.
     storage: memory
+    # Retention policy determines when messages should be deleted from the stream:
+    # (more info https://docs.nats.io/using-nats/developer/develop_jetstream/model_deep_dive#stream-limits-retention-and-policy):
+    # - interest: when all known observables have acknowledged a message it can be removed.
+    # - limits: messages are retained until any given limit is reached. This can be configured via maxMessages and maxBytes.
+    retentionPolicy: interest
```
Comment (raypinto) [inline resources/eventing/values.yaml:30]: Any reason to keep these configs here? The only reason the above two configs were moved from the `controller` values.yaml to the `eventing` values.yaml was to share these configs between the controller and the nats charts.

Comment (marcobebway) [reply resources/eventing/values.yaml]: Done.

**Review by raypinto — APPROVED** [review]

File: components/eventing-controller/pkg/env/nats_config.go [hunk 2]
```diff
@@ -48,6 +48,17 @@ type NatsConfig struct {
 	// Prefix for the JetStream stream subjects filter.
 	// It will be overridden by non-empty EventTypePrefix.
 	JSStreamSubjectPrefix string `envconfig:"JS_STREAM_SUBJECT_PREFIX" required:"true"`
+
+	// Deliver policy determines for a consumer where in the stream it wants to start receiving messages
```
Comment (NHingerl) [inline components/eventing-controller/pkg/env/nats_config.go:52]: `Deliver policy` is an awful word, but it's what NATS called it so I have to accept it. I'd format it in upper case so it's clear that it's a proper name.

```suggestion
	// Deliver Policy determines for a consumer where in the stream it starts receiving messages
```

I don't think we need the "want" ?

File: components/eventing-controller/pkg/env/nats_config.go [hunk 3]
```diff
@@ -48,6 +48,17 @@ type NatsConfig struct {
...
+	// (more info https://docs.nats.io/nats-concepts/jetstream/consumers#deliverpolicy-optstartseq-optstarttime):
+	// - all: The consumer will start receiving from the earliest available message.
```
Comment (NHingerl) [inline components/eventing-controller/pkg/env/nats_config.go:54]:
```suggestion
	// - all: The consumer starts receiving from the earliest available message.
```

File: components/eventing-controller/pkg/env/nats_config.go [hunk 4]
```diff
@@ -48,6 +48,17 @@ type NatsConfig struct {
...
+	// - last: When first consuming messages, the consumer will start receiving messages with the last message
```
Comment (NHingerl) [inline components/eventing-controller/pkg/env/nats_config.go:55]:
```suggestion
	// - last: When first consuming messages, the consumer starts receiving messages with the latest message.
```
I assume "last message" is the latest one?

File: components/eventing-controller/pkg/env/nats_config.go [hunk 5]
```diff
@@ -48,6 +48,17 @@ type NatsConfig struct {
...
+	//   added to the stream, so the very last message in the stream when the server realizes the consumer is ready.
```
Comment (NHingerl) [inline components/eventing-controller/pkg/env/nats_config.go:56]:
```suggestion
	//   added to the stream, so the very last message in the stream when the server realizes the consumer is ready.
```

I know you copied that from NATS but this whole sentence is extremely unclear to me. Can you explain in simpler terms? What does "so the very last message in the stream when the server realizes the consumer is ready." even mean?

File: components/eventing-controller/pkg/env/nats_config.go [hunk 6]
```diff
@@ -48,6 +48,17 @@ type NatsConfig struct {
...
+	// - last_per_subject: When first consuming messages, start with the latest one for each filtered subject
+	//   currently in the stream.
+	// - new: When first consuming messages, the consumer will only start receiving messages that were created
```
Comment (NHingerl) [inline components/eventing-controller/pkg/env/nats_config.go:59]:
```suggestion
	// - new: When first consuming messages, the consumer starts receiving messages that were created
```

File: resources/eventing/charts/controller/values.yaml [hunk 7]
```diff
@@ -80,9 +80,20 @@ jetstream:
   streamName: sap
   # Subject prefix for the JetStream stream when eventTypePrefix is not provided.
   subjectPrefix: kyma
-  # When should events be deleted from the stream:
-  # interest: when all known observables have acknowledged a message it can be removed.
-  # limits: messages are retained until any given limit is reached. This can be configured via maxMessages and maxBytes.
+  # Retention policy determines when messages should be deleted from the stream:
```
Comment (NHingerl) [inline resources/eventing/charts/controller/values.yaml:83]:
```suggestion
  # Retention policy determines when messages are deleted from the stream:
```

File: resources/eventing/charts/controller/values.yaml [hunk 8]
```diff
@@ -80,9 +80,20 @@ jetstream:
...
+  # (more info https://docs.nats.io/using-nats/developer/develop_jetstream/model_deep_dive#stream-limits-retention-and-policy):
+  # - interest: when all known observables have acknowledged a message it can be removed.
```
Comment (NHingerl) [inline resources/eventing/charts/controller/values.yaml:85]:
```suggestion
  # - interest: When all known observables have acknowledged a message, it can be removed.
```
Starting the sentence with upper case, just like in nats_config.go

File: resources/eventing/charts/controller/values.yaml [hunk 9]
```diff
@@ -80,9 +80,20 @@ jetstream:
...
+  # - limits: messages are retained until any given limit is reached. This can be configured via maxMessages and maxBytes.
```
Comment (NHingerl) [inline resources/eventing/charts/controller/values.yaml:86]:
```suggestion
  # - limits: Retain messages until any given limit is reached. Configure limits with maxMessages and maxBytes.
```
Active voice > passive voice

File: resources/eventing/charts/controller/values.yaml [hunk 10]
```diff
@@ -80,9 +80,20 @@ jetstream:
...
   retentionPolicy: interest
+  # Consumer deliver policy determines from where in the stream a consumer should start receiving messages
```
Comment (NHingerl) [inline resources/eventing/charts/controller/values.yaml:88]:
```suggestion
  # Consumer Deliver Policy determines from where in the stream a consumer starts receiving messages
```

File: resources/eventing/charts/controller/values.yaml [hunk 11]
```diff
@@ -80,9 +80,20 @@ jetstream:
...
+  # (more info https://docs.nats.io/nats-concepts/jetstream/consumers#deliverpolicy-optstartseq-optstarttime):
+  # - all: The consumer will start receiving from the earliest available message.
+  # - last: When first consuming messages, the consumer will start receiving messages with the last message
```
Comment (NHingerl) [inline resources/eventing/charts/controller/values.yaml:91]:
```suggestion
  # - last: When first consuming messages, the consumer starts receiving messages with the latest message.
```

File: components/eventing-controller/README.md [hunk 12]
```diff
@@ -71,6 +71,7 @@ This section explains how to use the Eventing Controller. It expects the followi
 |  `JS_STREAM_MAX_MSGS`             | The maximum number of messages in the stream. Used only when storage policy is set to `limits`. |
 |  `JS_STREAM_MAX_BYTES`            | The maximum size of the stream in bytes. Used only when storage policy is set to `limits`.     |
 |  `JS_STREAM_SUBJECT_PREFIX`       | The prefix for the JetStream subjects filter. Used only when `EVENT_TYPE_PREFIX` is empty.     |
+|  `JS_CONSUMER_DELIVER_POLICY`     | The policy to deliver events to consumers from the stream. Supported values are: `all`, `last`, `last_per_subject` and `new`. See https://docs.nats.io/nats-concepts/jetstream/consumers#deliverpolicy-optstartseq-optstarttime. |
```
[the remark this answers is not in the document]
Comment (marcobebway) [reply components/eventing-controller/README.md]: Done.

File: components/eventing-controller/pkg/env/nats_config.go (hunk 2 above)
Comment (marcobebway) [reply components/eventing-controller/pkg/env/nats_config.go]: Done.

File: components/eventing-controller/pkg/env/nats_config.go (hunk 3 above)
Comment (marcobebway) [reply components/eventing-controller/pkg/env/nats_config.go]: Done.

File: components/eventing-controller/pkg/env/nats_config.go (hunk 4 above)
Comment (marcobebway) [reply components/eventing-controller/pkg/env/nats_config.go]: Done.

File: components/eventing-controller/pkg/env/nats_config.go (hunk 5 above)
Comment (marcobebway) [reply components/eventing-controller/pkg/env/nats_config.go]: I removed it completely because the one  you suggested is sufficient.

File: components/eventing-controller/pkg/env/nats_config.go (hunk 6 above)
Comment (marcobebway) [reply components/eventing-controller/pkg/env/nats_config.go]: Done.

File: resources/eventing/charts/controller/values.yaml (hunk 8 above)
Comment (marcobebway) [reply resources/eventing/charts/controller/values.yaml]: Done.

File: resources/eventing/charts/controller/values.yaml (hunk 9 above)
Comment (marcobebway) [reply resources/eventing/charts/controller/values.yaml]: Done.

File: resources/eventing/charts/controller/values.yaml (hunk 10 above)
Comment (marcobebway) [reply resources/eventing/charts/controller/values.yaml]: Done.

File: resources/eventing/charts/controller/values.yaml [hunk 13]
```diff
@@ -80,9 +80,20 @@ jetstream:
...
-  # interest: when all known observables have acknowledged a message it can be removed.
-  # limits: messages are retained until any given limit is reached. This can be configured via maxMessages and maxBytes.
+  # Retention policy determines when messages should be deleted from the stream:
+  # (more info https://docs.nats.io/using-nats/developer/develop_jetstream/model_deep_dive#stream-limits-retention-and-policy):
+  # - interest: when all known observables have acknowledged a message it can be removed.
+  # - limits: messages are retained until any given limit is reached. This can be configured via maxMessages and maxBytes.
   retentionPolicy: interest
+  # Consumer deliver policy determines from where in the stream a consumer should start receiving messages
+  # (more info https://docs.nats.io/nats-concepts/jetstream/consumers#deliverpolicy-optstartseq-optstarttime):
+  # - all: The consumer will start receiving from the earliest available message.
```
[the remark this answers is not in the document]
Comment (marcobebway) [reply resources/eventing/charts/controller/values.yaml]: Done.

File: resources/eventing/charts/controller/values.yaml (hunk 11 above)
Comment (marcobebway) [reply resources/eventing/charts/controller/values.yaml]: Done.

File: components/eventing-controller/pkg/env/nats_config.go (hunk 5 above)
Comment (NHingerl) [reply components/eventing-controller/pkg/env/nats_config.go]: Nice!!

**Review by NHingerl — APPROVED** [review]

**Review by majakurcius — APPROVED** [review]
